### ISSUE TITLE
Rule change proposal regarding `prefer-ternary`

### ISSUE THREAD

**Issue by yeldiRium** [sindresorhus/eslint-plugin-unicorn#1079]
Ternaries are a useful syntactic construct for `const` assignments. They reduce a let/if/else mess to (hopefully) a single line. They can also be misused in various ways. The current implementation of the rule always enforces the use of a ternary if both branches of an if/else expression return the same type. This is unfortunately coincidentally true for Promises and Errors. Take these two examples:

```javascript
if (condition) {
  await doSomething();
} else {
  await doSomeOtherThing();
}
```
The rule corrects this to:
```javascript
await (condition ? doSomething() : doSomeOtherThing());
```

```javascript
if (condition) {
  throw new ThisKindOfError();
} else {
  throw new ThatKindOfError();
}
```
The rule corrects this to:
```javascript
throw (condition ? new ThisKindOfError() : new ThatKindOfError());
```

Both cases I would categorize as misuse of ternaries, since they do nothing more than deduplicate a `await`/`throw` and make the code less readable.

However, I would still like the rule to catch valid cases, which mostly come in this form:

```javascript
let someVariable;

if (condition) {
  someVariable = thisValue;
} else {
  someVariable = otherValue;
}
```
This is (correctly) corrected to:
```javascript
const someVariable = condition ? thisValue : otherValue;
```

I propose that the rule should be configurable to only apply to assignments. This would exclude situations using await/throw/yield.

**Comment by fisker** [comment]
Currently, we support 5 kinds of AST to merge, and we support nested

```js
// This is fixed to `return await (b ? c : d)`
async function a() {
	if (b) {
		return await c;
	} else {
		return await d;
	}
}
```

We already support `"always"/"only-single-line"`, so options will like

```js
{
	'unicorn/prefer-ternary': [
		'error',
		'always',
		{
			return: true,
			yield: true, // Include `yield*`
			await: true,
			throw: true,
			assignment: true
		}
	]
}
```

Do we need merge this two together? Like

```js
{
	'unicorn/prefer-ternary': [
		'error',
		{
			onlySingleLine: false,
			return: true,
			yield: true, // Include `yield*`
			await: true,
			throw: true,
			assignment: true
		}
	]
}
```

**Comment by yeldiRium** [comment]
I don't know what the best practices on eslint-rule-configuration are, but I think the first solution looks good. It is non-breaking and it distinguishes the more detail-oriented options from the broad-behavior-influencing one.

**Comment by sindresorhus** [comment]
Instead of:

```js
await (condition ? doSomething() : doSomeOtherThing());
```

You should do:

```js
const promise = condition ? doSomething() : doSomeOtherThing();
await promise;
```

**Comment by yeldiRium** [comment]
Regarding

```typescript
const promise = condition ? doSomething() : doSomeOtherThing();
await promise;
```

I agree that it is better than what the rule suggests, but it is still much worse than

```typesript
if (condition) {
  await doSomething();
} else {
  await doSomeOtherThing();
}
```

since it conceals the side-effects of those asynchronous operations.

I am for the suggested more options, since in our case we prefer assignments and returns to be checked. If there are no other voices on this, a single option that only checks assignments and returns would be fine in my opinion. But I can see that others might only want assignments or even only returns.

**Comment by fregante** [comment]
> Instead of:
> 
> ```js
> await (condition ? doSomething() : doSomeOtherThing());
> ```
> 
> You should do:
> 
> ```js
> const promise = condition ? doSomething() : doSomeOtherThing();
> await promise;
> ```

I really don't see how that's better than just using a regular if/else 😟 

```js
if (condition) {
  await doSomething();
} else {
  await doSomeOtherThing();
}
```

I just found my code fixed from:

```js
if (shouldUpdate) {
	await updateDeployment(process.env.SMART_DEPLOYMENT_ID);
} else {
	await createDeployment();
}
```

to


```js
await (shouldUpdate ? updateDeployment(process.env.SMART_DEPLOYMENT_ID) : createDeployment());
```

and I don’t think this would be any better, it's even longer just to avoid 2 parens


```js
const promise = shouldUpdate ? updateDeployment(process.env.SMART_DEPLOYMENT_ID) : createDeployment();
await promise
```

**Comment by sindresorhus** [comment]
> const promise = shouldUpdate ? updateDeployment(process.env.SMART_DEPLOYMENT_ID) : createDeployment();
await promise

Yeah, I agree. It seemed nice at first but having used it in practice, I agree, an if-statement would be better here.

**Comment by RedGuy12** [comment]
Bumping this issue. I would really appreciate an option to only prefer ternaries in assignments and `return` statements. If this is accepted, I'm willing to try adding this.

*Sidenote: I believe #1633 and #867 are the same as this issue?*

**Comment by RedGuy12** [comment]
@fregante @sindresorhus apologies for ping, is there a way to get an answer on this issue? is making a PR the best way?

**Comment by sindresorhus** [comment]
A pull request to add an option for this would be welcome. I think it should be enabled by default (breaking change).

**Comment by yeldiRium** [comment]
Closing, since I'm cleaning up my open issues and don't care about this anymore. If anyone still wants this, please re-open.
